### PR TITLE
dd-7-search-form

### PR DESCRIPTION
**Pull request by ChickenMcNuggetss**
# Task

[Search form](https://divas801801.atlassian.net/jira/software/projects/DD/boards/2?selectedIssue=DD-7)

# Description

Added search form

## Pull request type

Please check the type of change your PR introduces:

- [x] 🍕 Feature
- [ ] 🐛 Bugfix
- [ ] 🧑‍💻 Code Refactor
- [ ] 📝 Documentation Update
- [ ] 🤖 Build related changes
- [ ] 🔥 Performance Improvements
- [ ] Other (please describe):

# Demo (situationally)

# Has It Been Tested?

- [x] Yes
- [ ] No, and this is why: please replace this line with details on why tests have not been included
- [ ] Tests are not needed for this task
- [ ] I need help with writing tests

### CODE REVIEW

File: src/app/features/search/services/search/search.service.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  private httpClient = inject(HttpClient);
+
+  search({ fromLatitude, fromLongitude, toLatitude, toLongitude, time }: SearchRouteParams) {
+    const params = {
+      fromLatitude,
+      fromLongitude,
+      toLatitude,
+      toLongitude,
+      time,
+    };
```
Comment (TELEUZI) [inline src/app/features/search/services/search/search.service.ts]: Why do you need object destructuring?

File: src/app/features/search/services/search/search.service.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    return this.httpClient.get('/api/search', { params }).pipe(catchError((error) => of(error)));
```
Comment (TELEUZI) [inline src/app/features/search/services/search/search.service.ts]: `catchError` is useless here, we should return some real value in it, please check https://dev.to/georgeknap/the-catch-in-the-rxjs-catcherror-51mp

Comment (TELEUZI) [reply src/app/features/search/services/search/search.service.ts]: https://thabo-ambrose.medium.com/differentiating-between-catcherror-throwerror-new-error-and-throw-new-error-94555ec62a53

File: src/app/features/search/services/location-api/location-api.service.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+const BASE_URL = 'http://api.openweathermap.org/geo/1.0';
+const responseLimit = 5;
+
+@Injectable({
+  providedIn: 'root',
+})
+export class LocationApiService {
+  private httpClient = inject(HttpClient);
+
+  apiLocationKey = import.meta.env.NG_APP_API_LOCATION_KEY;
```
Comment (TELEUZI) [inline src/app/features/search/services/location-api/location-api.service.ts]: Use `environment.ts`

File: src/app/features/search/services/location-api/location-api.service.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  getLocationCoordinates(city: string) {
```
Comment (TELEUZI) [inline src/app/features/search/services/location-api/location-api.service.ts]: Modifier

File: src/app/features/search/services/location-api/location-api.service.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import { CityInfo } from '../../interfaces/city-info';
+
+const BASE_URL = 'http://api.openweathermap.org/geo/1.0';
+const responseLimit = 5;
+
+@Injectable({
+  providedIn: 'root',
+})
+export class LocationApiService {
+  private httpClient = inject(HttpClient);
```
Comment (TELEUZI) [inline src/app/features/search/services/location-api/location-api.service.ts]: `readonly`

File: src/app/features/search/services/location-api/location-api.service.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { HttpClient } from '@angular/common/http';
+import { inject, Injectable } from '@angular/core';
+import { CityInfo } from '../../interfaces/city-info';
+
+const BASE_URL = 'http://api.openweathermap.org/geo/1.0';
```
Comment (TELEUZI) [inline src/app/features/search/services/location-api/location-api.service.ts]: Put this url with more specific name into `environment.ts`

File: src/app/features/search/interfaces/search-route-response.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  rideId: number;
+  segments: Array<RoadSection>;
+}
+
+interface RoadSection {
+  time: Array<string>;
+  price: {
+    [carriageType: string]: number;
+  };
+  occupiedSeats: Array<number>;
```
Comment (TELEUZI) [inline src/app/features/search/interfaces/search-route-response.ts]: `number[]`

File: src/app/features/search/interfaces/search-route-response.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  carriages: Array<string>;
+  schedule: Array<Ride>;
+}
+
+interface Ride {
+  rideId: number;
+  segments: Array<RoadSection>;
+}
+
+interface RoadSection {
```
Comment (TELEUZI) [inline src/app/features/search/interfaces/search-route-response.ts:33]: `export`?

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const searchServiceResponse = this.searchService
+      .search({
+        fromLatitude: this.fromInputValue.lat,
+        fromLongitude: this.fromInputValue.lon,
+        toLatitude: this.toInputValue.lat,
+        toLongitude: this.toInputValue.lon,
+        time: date,
+      })
+      .subscribe();
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: ` this.subscription.add`

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      this.to.setValue(city.name);
+    }
+  }
+
+  search() {
+    if (this.searchForm.invalid) return;
+    let date = 0;
+    if (this.date.value) {
+      const { day, month, year } = this.date.value[0];
+      const { hours, minutes } = this.date.value[1];
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Use array destructuring

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+      })
+    );
+  }
+
+  updateCities(city: string, inputName: string) {
+    this.locationService.getLocationCoordinates(city).subscribe((value) => {
+      if (inputName === 'from') {
+        this.fromInputValue = { lat: value[0].lat, lon: value[0].lon };
+      } else {
+        this.toInputValue = { lat: value[0].lat, lon: value[0].lon };
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts:74]: `value[1].lat`?

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        if (!this.from.value) return;
+        this.updateCities(this.from.value, 'from');
+        if (!this.to.value) return;
+        this.updateCities(this.to.value, 'to');
+      })
+    );
+  }
+
+  updateCities(city: string, inputName: string) {
+    this.locationService.getLocationCoordinates(city).subscribe((value) => {
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: `value`?

File: src/app/features/search/components/search-form/search-form.component.html [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  <tui-input *tuiLet="cities$() as cities" formControlName="to" class="search-input">
+    To
+    <input placeholder="Type city name" tuiTextfieldLegacy />
+    <ng-container *ngIf="cities?.length">
+      <tui-data-list *tuiDataList>
+        <button *ngFor="let city of cities" tuiOption [value]="city.name" (click)="onSelected(city, 'to')">
+          {{ city.name }}, {{ city.country }}
+        </button>
+      </tui-data-list>
+    </ng-container>
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.html]: Try using ng-template and pass `to` or `from` as context: 
https://www.youtube.com/watch?v=vfPVdJ2oQlM
https://www.tektutorialshub.com/angular/ngtemplateoutlet-in-angular/
https://blog.angular-university.io/angular-ng-template-ng-container-ngtemplateoutlet/

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+    from: new FormControl('', [Validators.required]),
+    to: new FormControl('', [Validators.required]),
+    date: new FormControl<[TuiDay, TuiTime]>([new TuiDay(2017, 2, 15), new TuiTime(12, 30)]),
+  });
+
+  ngOnInit(): void {
+    this.subscription.add(
+      merge(
+        this.searchForm.controls.from.valueChanges.pipe(debounceTime(1000)),
+        this.searchForm.controls.to.valueChanges.pipe(debounceTime(1000))
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]:
```suggestion
        const {from, to} = this.searchForm.controls;
        from.valueChanges.pipe(debounceTime(1000)),
        to.valueChanges.pipe(debounceTime(1000))
```

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+      ).subscribe(() => {
+        if (!this.from.value) return;
+        this.updateCities(this.from.value, 'from');
+        if (!this.to.value) return;
+        this.updateCities(this.to.value, 'to');
+      })
+    );
+  }
+
+  updateCities(city: string, inputName: string) {
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Looks like `inputName: string` is not any string, but some concreate ones

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+export class SearchFormComponent implements OnInit, OnDestroy {
+  private searchService = inject(SearchService);
+
+  private locationService = inject(LocationApiService);
+
+  private subscription = new Subscription();
+
+  cities$ = this.searchService.cities;
+
+  private fromInputValue: { lat: number; lon: number } = { lat: 0, lon: 0 };
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: { lat: number; lon: number } looks like some interface

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+      merge(
+        this.searchForm.controls.from.valueChanges.pipe(debounceTime(1000)),
+        this.searchForm.controls.to.valueChanges.pipe(debounceTime(1000))
+      ).subscribe(() => {
+        if (!this.from.value) return;
+        this.updateCities(this.from.value, 'from');
+        if (!this.to.value) return;
+        this.updateCities(this.to.value, 'to');
+      })
+    );
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Looks like we should separate those streams.

File: src/app/features/search/components/search-form/search-form.component.ts (hunk 11 above)
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Use destructuring on `value[0]`, please

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+      }
+      this.searchService.setCities(value);
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.subscription.unsubscribe();
+  }
+
+  onSelected(city: CityInfo, inputName: string) {
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Looks like `inputName: string` is not any string, but some concreate ones

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    if (inputName === 'to') {
+      this.to.setValue(city.name);
+    }
+  }
+
+  search() {
+    if (this.searchForm.invalid) return;
+    let date = 0;
+    if (this.date.value) {
+      const { day, month, year } = this.date.value[0];
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Date validation?

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+  cities$ = this.searchService.cities;
+
+  private fromInputValue: { lat: number; lon: number } = { lat: 0, lon: 0 };
+
+  private toInputValue: { lat: number; lon: number } = { lat: 0, lon: 0 };
+
+  searchForm = new FormGroup({
+    from: new FormControl('', [Validators.required]),
+    to: new FormControl('', [Validators.required]),
+    date: new FormControl<[TuiDay, TuiTime]>([new TuiDay(2017, 2, 15), new TuiTime(12, 30)]),
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Let's set current/tomorrow date as default value

Comment (TELEUZI) [reply src/app/features/search/components/search-form/search-form.component.ts]: We could also not set time by default

File: src/app/features/search/components/search-form/search-form.component.ts [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SearchFormComponent implements OnInit, OnDestroy {
+  private searchService = inject(SearchService);
+
+  private locationService = inject(LocationApiService);
+
+  private subscription = new Subscription();
+
+  cities$ = this.searchService.cities;
```
Comment (TELEUZI) [inline src/app/features/search/components/search-form/search-form.component.ts]: Why `$` here?

File: src/app/features/search/services/search/search.service.ts (hunk 2 above)
Comment (ChickenMcNuggetss) [reply src/app/features/search/services/search/search.service.ts]: Thanks for materials

**Review by TELEUZI — DISMISSED** [review]

File: .eslintrc.json [hunk 22]
```diff
@@ -43,6 +43,7 @@
             ]
           }
         ],
+        "no-underscore-dangle": "off",
```
Comment (RykovaMariia) [inline .eslintrc.json]: this rule is already on line 29

File: package-lock.json
Comment (RykovaMariia) [inline package-lock.json:1]: why?

File: src/app/app.component.scss [hunk 23]
```diff
@@ -1,4 +1,7 @@
 :host {
+  max-width: 1440px;
+  margin: 0 auto;
+  padding: 0 1rem;
```
Comment (RykovaMariia) [inline src/app/app.component.scss]: this is already on line 8-11

File: src/app/app.component.scss [hunk 24]
```diff
@@ -12,3 +15,9 @@ main {
   display: flex;
   justify-content: center;
 }
+
+main {
+  display: flex;
+  justify-content: center;
+  width: 100%;
+}
```
Comment (RykovaMariia) [inline src/app/app.component.scss]: this is already on line 14-17

File: src/app/app.routes.ts [hunk 25]
```diff
@@ -7,7 +7,7 @@ export const routes: Routes = [
   {
     path: 'home',
     loadComponent: () =>
-      import('@features/TrainA/pages/home-page/home-page.component').then((m) => m.HomePageComponent),
+      import('@features/search/pages/search-page/search-page.component').then((m) => m.SearchPageComponent),
```
Comment (RykovaMariia) [inline src/app/app.routes.ts]: why search?

File: src/environment.ts [hunk 26]
```diff
@@ -0,0 +1,4 @@
+export const environment = {
+  baseLocationApiUrl: 'http://api.openweathermap.org/geo/1.0',
+  locationApiKey: import.meta.env.NG_APP_API_LOCATION_KEY,
+};
```
Comment (RykovaMariia) [inline src/environment.ts]: `as const ` ?

File: src/app/features/search/components/search-form/search-form.component.html [hunk 27]
```diff
@@ -0,0 +1,25 @@
+<form class="container" [formGroup]="searchForm" (ngSubmit)="search()">
+  <ng-container [ngTemplateOutlet]="input" [ngTemplateOutletContext]="{ title: 'from' }"></ng-container>
+  <ng-container [ngTemplateOutlet]="input" [ngTemplateOutletContext]="{ title: 'to' }"></ng-container>
+  <tui-input-date-time formControlName="date" timeMode="HH:MM"> Choose date and time </tui-input-date-time>
+  <button tuiButton appearance="primary" size="m" type="submit">Search</button>
+</form>
+
+<ng-template #input let-title="title">
+  <tui-input *tuiLet="cities() as cities" [formControl]="$any(searchForm.controls)[title]" class="search-input">
```
Comment (Dewlar) [inline src/app/features/search/components/search-form/search-form.component.html]: can we use a standard `tui` classes for responsibility design ? like this?
image
image
image
```
@import '@taiga-ui/core/styles/taiga-ui-local.scss';
.submit {
  flex: 0 1 10rem;

  @media #{$tui-mobile} {
  flex: 1 1 10rem;
}
}
```

File: .eslintrc.json (hunk 22 above)
Comment (ChickenMcNuggetss) [reply .eslintrc.json]: Thanks!

File: src/app/app.routes.ts (hunk 25 above)
Comment (ChickenMcNuggetss) [reply src/app/app.routes.ts]: In that case, I'll rename it to:
```suggestion
      import('@features/home/pages/home-page/home-page.component').then((m) => m.HomePageComponent),
```

File: src/app/features/search/components/search-form/search-form.component.html (hunk 27 above)
Comment (ChickenMcNuggetss) [reply src/app/features/search/components/search-form/search-form.component.html]: Thank you so much!

File: src/app/features/home/components/search-form/search-form.component.ts [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+  private fromCityCoordinates: CityCoordinates = { latitude: 0, longitude: 0 };
+
+  private toCityCoordinates: CityCoordinates = { latitude: 0, longitude: 0 };
+
+  private selectedCityIndex: number = 0;
+
+  public searchForm = new FormGroup({
+    from: new FormControl('', [Validators.required]),
+    to: new FormControl('', [Validators.required]),
+    date: new FormControl<[TuiDay, TuiTime]>([TuiDay.currentUtc(), new TuiTime(0, 0)], dateValidator()),
```
Comment (YuliyaVoronovich) [inline src/app/features/home/components/search-form/search-form.component.ts]: use FormBuilder please)

**Review by YuliyaVoronovich — DISMISSED** [review]

**Review by RykovaMariia — APPROVED** [review]

**Review by YuliyaVoronovich — APPROVED** [review]

**Pull request merged**